chain/src/chain.rs

```diff
@@ -1300,12 +1299,40 @@ fn setup_head(
 ) -> Result<(), Error> {
 	let mut batch = store.batch()?;
 
+	// Take existing chain tip as a starting point and use it to find the first valid
+	// header in our db. Go back through the chain by height until we find a valid header.
+	// The tip for this valid header is our head.
+	// This ensures we handle the hardfork scenario where our node processed what appeared to be a
+	// valid header but post-hardfork is no longer valid (and must be rewound).
+	// Specifically the header version number relative to block height.
+	let head = if let Ok(head) = batch.head() {
```

Comment:
I don't know what all `head` is used for in this context, but this isn't actually rewinding the txhashset, so that could be problematic.

Reply:
It looks like it is done below, so this should be good. https://github.com/mimblewimble/grin/blob/737932ccb021560a0de72d997af4ba2b16e77d5f/chain/src/chain.rs#L1359

Reply:
Yeah - step 1 is "find a valid header", step 2 is "rewind txhashsets to find a valid chain state".